File: 2023/day01.py
# Advent of Code - 2023
## Day 1

import re

# Part 1

def find_digit_part1(line):
    """
    Find a digit by checking if it is int()able.
    :param line: string containing numbers and letetrs
    :return: found digit as string
    """
    for character in line: 
        try:
            digit = int(character)
            return character
        except ValueError:
            continue

def part1(lines):
    """
    Executing the finding of digits for a list of strings from the start, 
    then from the end.
    Adding both digits together to create a 2-digit number 
    and then add them up for all lines.
    :param lines: a list of strings
    :return: sum of all numbers found
    """
    sum = 0

    for line in lines: 
        first_digit = find_digit_part1(line)
        second_digit = find_digit_part1(line[::-1])
        number = first_digit + second_digit
        sum += int(number)
    return sum

# Part 2

# Preparation of containers
conversion = {
    'one': 1, 
    'two': 2, 
    'three': 3, 
    'four': 4,
    'five': 5,
    'six': 6, 
    'seven': 7, 
    'eight': 8,
    'nine': 9
}
    
written_digits = [digit for digit, number in conversion.items()]
actual_digits = [str(x+1) for x in range(9)]

def find_substring(line):
    """
    Find all digits that are written out in a string.
    :param line: string containing letters and numbers
    :return: list of tuples (position of match, match)
    """
    found_digits = []
    for written_digit in written_digits:
        if written_digit in line:
            for match in re.finditer(written_digit, line):
                found_digits.append((match.start(), written_digit))
    return found_digits

def find_digit_part2(line):
    """
    Find all digits in a string.
    :param line: string containing letters and numbers
    :return: list of tuples (position of match, match)
    """
    found_numbers=[]
    for actual_digit in actual_digits:
        if actual_digit in line:
            for match in re.finditer(actual_digit, line):
                found_numbers.append((match.start(), actual_digit))
    return found_numbers


def find_most_left_and_most_right_digit(line):
    """
    Find the most left and most right digit found in a string.
    :param line: string
    :return: return tuple of two digits (either word or number)
    """
    substrings = find_substring(line)
    numbers = find_digit_part2(line)
    digits = substrings + numbers
    sorted_digits = sorted(digits)
    return sorted_digits[0][1], sorted_digits[-1][1]

def convert_word_to_digit(digit):
    """
    Converts strings that represent a digit to a digit.
    :param digit: a digit representation either word or number
    :return: string
    """
    try: 
        result = int(digit)
    except ValueError:
        result = conversion[digit]
    return str(result)


def part2(lines):
    """
    Execute part two: find digits and convert them to numbers.
    Then add the numbers up - digits can be words or numbers.
    :param: list of strings
    :return: sum of found numbers
    """
    sum = 0
    for line in lines:
        first, last = find_most_left_and_most_right_digit(line)
        first_digit = convert_word_to_digit(first)
        second_digit = convert_word_to_digit(last)
        number = first_digit + second_digit
        sum += int(number)
    return sum


# with open('puzzle_input') as file:
with open('test_input', 'r') as file:
    lines = file.readlines()

print("- -      -     -   *  -    -     -      -  *  *  - -   ")
print("*   -    .   .    .       *     .  .   .    *       -  ")
print(f"{'Advent of Code 2023 - Day 1':^55}")
print(".       .      *      -        -     *     .     .    .")
print("    -      .    -  *    -    -    *    .  .  .    *   -")
print(f"Part 1: {part1(lines):^55}")
print(f"Part 2: {part2(lines):^55}")
print("    -      .    -  *    -    -    *    .  .  .    *   -")
print(".       .      *      -        -     *     .     .    .")

File: 2023/day01.cpp
/** Advent of Code - 2024
 * Day 01
 **/

#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <ctype.h>
using namespace std;

vector<string> WRITTENDIGITS = { "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };

vector<string> getLines(int lines, string filename)
{
    vector<string> puzzle;
    string temp;
    ifstream f(filename);
    while (getline(f, temp))
        puzzle.push_back(temp);
    return puzzle;
}

char getInteger(string& line)
{
    for (auto l : line)
    {
        if (isdigit(l))
        {
            return l;
        }
    }
}

void part1(vector<string>& puzzle)
{
    int calibrationSum = 0;
    for (string line : puzzle)
    {
        //cout << line << endl;
        char a = getInteger(line);
        string reverse = string(line.rbegin(), line.rend());
        char b = getInteger(reverse);
        string c = string() + a + b;
        int d = stoi(c);
        //cout << d << endl;
        calibrationSum += d;
    }

    cout << "Part 1: " << calibrationSum << endl;
}

void part2(vector<string>& puzzle)
{
    int part2sum = 0;

    for (string line : puzzle)
    {
        int smallestindex = line.length();
        int biggestindex = -1;
        string smallestdigit = "";
        string biggestdigit = "";
        int i = 1;

        for (int i = 0; i < 10; i++)
        {
            vector<pair<int, int>> bla;
            std::string::size_type pos = 0;
            for (;;)
            {
                pos = line.find(WRITTENDIGITS[i], pos);
                if (pos == line.npos)
                    break;
                bla.push_back(pair<int, int>(pos, i));
                ++pos;
            }
            for (auto b : bla)
            {
                if (b.first < smallestindex)
                {
                    smallestindex = b.first;
                    smallestdigit = to_string(b.second);
                }
                if (b.first > biggestindex)
                {
                    biggestindex = b.first;
                    biggestdigit = to_string(b.second);
                }
            }
        }

        for (int j = 1; j < 10; j++)
        {
            vector<pair<int, int>> bla;
            std::string::size_type pos = 0;
            for (;;)
            {
                pos = line.find(to_string(j), pos);
                if (pos == line.npos)
                    break;
                bla.push_back(pair<int, int>(pos, j));
                ++pos;
            }
            for (auto b : bla)
            {
                if (b.first < smallestindex)
                {
                    smallestindex = b.first;
                    smallestdigit = to_string(b.second);
                }
                if (b.first > biggestindex)
                {
                    biggestindex = b.first;
                    biggestdigit = to_string(b.second);
                }
            }
        }
        part2sum += stoi(smallestdigit + biggestdigit);
    }

    cout << "Part 2: " << part2sum << endl;
}

int main()
{
    // READING FROM FILE
    const int lines = 4;
    vector<string> puzzle = getLines(lines, "puzzle.txt");

    cout << "- -      -     -   *  -    -     -      -  *  *  - -   " << endl;
    cout << "*   -    .   .    .       *     .  .   .    *       -  " << endl;
    cout << "Advent of Code 2023 - Day 1" << endl;
    part1(puzzle);
    part2(puzzle);
    cout << "    -      .    -  *    -    -    *    .  .  .    *   -" << endl;
    cout << ".       .      *      -        -     *     .     .    ." << endl;

    return 0;
}
